src/main/java/com/commercetools/sync/sdk2/states/helpers/StateReferenceResolver.java

```diff
@@ -0,0 +1,139 @@
+   */
+  @Override
+  @Nonnull
+  public CompletionStage<StateDraft> resolveReferences(@Nonnull final StateDraft stateDraft) {
+    return resolveTransitionReferences(StateDraftBuilder.of(stateDraft))
+        .thenApply(StateDraftBuilder::build);
+  }
+
+  /**
+   * Given a {@link StateDraftBuilder} this method attempts to resolve the transitions to return a
+   * {@link java.util.concurrent.CompletionStage} which contains a new instance of the builder with
+   * the resolved references. The key of the state references is taken from the value of the id
+   * fields.
+   *
+   * @param draftBuilder the state to resolve its transition references.
+   * @return a {@link java.util.concurrent.CompletionStage} that contains as a result a new builder
+   *     instance with resolved references or, in case an error occurs during reference resolution,
+   *     a {@link ReferenceResolutionException}.
+   */
+  @Nonnull
+  private CompletionStage<StateDraftBuilder> resolveTransitionReferences(
+      @Nonnull final StateDraftBuilder draftBuilder) {
+
+    final List<StateResourceIdentifier> stateReferences = draftBuilder.getTransitions();
+    final Set<String> stateKeys = new HashSet<>();
+    if (stateReferences != null) {
+      for (StateResourceIdentifier stateReference : stateReferences) {
+        if (stateReference != null) {
+          try {
+            final String stateKey = getKeyFromResourceIdentifier(stateReference);
+            stateKeys.add(stateKey);
+          } catch (ReferenceResolutionException referenceResolutionException) {
+            return exceptionallyCompletedFuture(
+                new ReferenceResolutionException(
+                    format(
+                        FAILED_TO_RESOLVE_REFERENCE,
+                        draftBuilder.getKey(),
+                        referenceResolutionException.getMessage())));
+          }
+        }
+      }
+    }
+    return fetchAndResolveStateTransitions(draftBuilder, stateKeys);
+  }
+
+  /**
+   * Given a {@link StateDraftBuilder} and a {@link java.util.Set} of {@code stateKeys} this method
+   * fetches the states corresponding to these keys. Then it sets the state references on the {@code
+   * draftBuilder}.
+   *
+   * @param draftBuilder the state draft builder to resolve it's state references.
+   * @param stateKeys the state keys of to resolve their actual id on the draft.
+   * @return a {@link java.util.concurrent.CompletionStage} that contains as a result a new
+   *     stateDraft instance with resolved state resource identifiers or an exception.
+   */
+  @Nonnull
+  private CompletionStage<StateDraftBuilder> fetchAndResolveStateTransitions(
+      @Nonnull final StateDraftBuilder draftBuilder, @Nonnull final Set<String> stateKeys) {
+
+    return stateService
+        .fetchMatchingStatesByKeysWithTransitions(stateKeys)
+        .thenApply(
+            states ->
+                states.stream().map(State::toReference).filter(Objects::nonNull).collect(toSet()))
```

Comment:
Why do you convert `toReference` first? Why not `toStateResourceIdentifier` at this point?

Reply:
https://github.com/commercetools/commercetools-sync-java/pull/1033/commits/b65043df28d62287c2bf9236d746dba4eed9bb0e